### PR TITLE
Various additions

### PR DESCRIPTION
**Pull request by distributive**
**Commands**
- `say` - outputs its input
- `favourite` - generates a category of thing
- `flip` - flips a coin, or randomly selects from a list

**Helper plugins**
- `Util.Error.hs` - handles custom error types (note I plan to overhaul this very soon)
- `Util.Random.hs` - handles random functions

**Parser functions**
- `notSpace` - matches any character that isn't a space
- `nonSpaceWord` - matches a word, but allows any non-space character in the word
- `skipSpace1` - skips one or more spaces, but not zero
- `untilEnd` - no longer requires at least one character
- `untilEnd1` - does what `untilEnd` used to (matched the rest of an input, requiring one character or more)

**Fixes**
- Commands now expect a space before their arguments
- `quote add` now requires an author (previously accepted whitespace)

### CODE REVIEW

File: src/Tablebot/Plugins/Welcome.hs
Comment (L0neGamer) [inline src/Tablebot/Plugins/Welcome.hs:40]: the name of the command is not say (see long help text)

File: src/Tablebot/Util/Random.hs
Comment (L0neGamer) [inline src/Tablebot/Util/Random.hs:38]: It may be worth explaining this line a lil bit because it's a bit loaded

File: src/Tablebot/Plugins/Say.hs
Comment (L0neGamer) [inline src/Tablebot/Plugins/Say.hs:25]: Correct me if I'm wrong but does this make the error message just be blank if it can't find any items? is the `<?>` needed?

File: src/Tablebot/Util/Random.hs
Comment (L0neGamer) [inline src/Tablebot/Util/Random.hs:22]: I've made my opinions on `!!` known - and would support using `chooseOneWeighted` under the hood even if it has those extra checks

File: src/Tablebot/Util/Random.hs
Comment (distributive) [reply src/Tablebot/Util/Random.hs:38]: Understanding code is left as an exercise for the maintainer.

Comment (distributive) [reply src/Tablebot/Util/Random.hs:40]: Added.

File: src/Tablebot/Util/Random.hs
Comment (finnbar) [reply src/Tablebot/Util/Random.hs:22]: I don't like it, but agree that this is likely the best way to do it so I have no problems.

**Review by finnbar — CHANGES_REQUESTED** [review]
Approved once the fixes are made.

File: src/Tablebot/Plugins/Welcome.hs
Comment (distributive) [reply src/Tablebot/Plugins/Welcome.hs:74]: As discussed I'm leaving this as it is for now and will raise an issue. I'll have a look into it at a later date.

**Pull request merged**
